### PR TITLE
Implement tiling patterns for the SVG back-end

### LINKED ISSUE #10796: When converting to SVG the "fill-opacity: 1" attribute should not be added if the "fill" attribute is missing
Attach (recommended) or Link to PDF file here:
 [http://www.mediafire.com/file/jz26jy0zz326pw3/hypothermia-prevention-and-management-in-adults-having-surgery-update+Dec2016.pdf](http://www.mediafire.com/file/jz26jy0zz326pw3/hypothermia-prevention-and-management-in-adults-having-surgery-update+Dec2016.pdf)

Configuration:
- Web browser and its version: Google Chrome 73.0.3683.103 (Official Build) (64-bit)
- Operating system and its version: Ubuntu 18.04
- PDF.js version: "pdfjs-dist": "~2.0.943"
- Is a browser extension: No

Steps to reproduce the problem:
1. Upload and convert the pdf file to SVG.

What is the expected behavior?
Create the svg file without the "fill-opacity: 1" attribute since the "fill" attribute is missing.

What went wrong? (add screenshot)
The pages from 2 to the last have the "fill-opacity: 1" attribute, but do not have the "fill" attribute, so the browser paints all these pages into black color by default: [https://www.dropbox.com/s/fgrcdr5573cwf81/Screen%20Shot%202019-05-02%20at%2015.37.21.png?dl=0](https://www.dropbox.com/s/fgrcdr5573cwf81/Screen%20Shot%202019-05-02%20at%2015.37.21.png?dl=0)

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): None

### PR DESCRIPTION
**Pull request by ahuglajbclajep**
It may not be perfect, but some PDFs, such as #10796, can be rendered.

Fixes #10796.
Improves #4985.

### CODE REVIEW

File: src/display/svg.js [hunk 1]
```diff
@@ -982,8 +982,54 @@ SVGGraphics = class SVGGraphics {
...
+      const strokeColor = this.current.strokeColor;
+
+      const bbox = this.svgFactory.create(tx1 - tx0, ty1 - ty0);
+      this.svg = bbox;
+      this.transformMatrix = matrix;
+      if (paintType === 2) {
+        const cssColor = Util.makeCssRgb(...color);
+        this.current.fillColor = cssColor;
+        this.current.strokeColor = cssColor;
+      }
```
Comment (timvandermeij) [inline src/display/svg.js]: Let's change this to:

```javascript
} else {
  warn(`Unimplemented tiling type ${paintType}`);
}
```

This way we still get notified in the console if an unsupported tiling type is used, which helps for debugging issues.

File: src/display/svg.js [hunk 2]
```diff
@@ -982,8 +982,54 @@ SVGGraphics = class SVGGraphics {
    */
   _makeColorN_Pattern(args) {
     if (args[0] === 'TilingPattern') {
-      warn('Unimplemented pattern TilingPattern');
-      return null;
+      const color = args[1];
```
Comment (timvandermeij) [inline src/display/svg.js]: Perhaps it's better to split the logic off to a new method named `_makeTilingPattern`, just like `_makeShadingPattern` below, for consistency.

**Review by timvandermeij — COMMENTED** [review]
I haven't tested this yet, but I really like that the implementation is short and easy to follow!

> It may not be perfect, [...]

That's perfectly okay, in fact, it's much easier to review/merge if the patch is small. It's always preferred to have multiple smaller iterations than one big patch.

File: src/display/svg.js (hunk 1 above)
Comment (ahuglajbclajep) [reply src/display/svg.js]: paintType can be only `1` or `2` in the [PDF specification](https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdf_reference_archive/pdf_reference_1-7.pdf#page=292).
This implementation also includes processing when `paintType === 1`, so its `else` clause seems unnecessary.

Comment (timvandermeij) [reply src/display/svg.js]: You're right. Thank you for clarifying this.

File: src/display/svg.js [hunk 3]
```diff
@@ -982,12 +982,65 @@ SVGGraphics = class SVGGraphics {
...
+    }
+    this.executeOpTree(this.convertOpList(operatorList));
+
+    // Restore saved state.
+    this.svg = svg;
+    this.transformMatrix = transformMatrix;
+    this.fillColor = fillColor;
+    this.strokeColor = strokeColor;
+
+    tiling.appendChild(bbox.childNodes[0]);
```
Comment (timvandermeij) [inline src/display/svg.js]: This line appears to break the PDF file from #10487, which also uses tiling patterns. Could you check why that happens?

**Review by timvandermeij — COMMENTED** [review]
I think this is good to merge when the final comment is addressed. I have updated the title and the description of the pull request with additional details and which issues are fixed/improved by this patch.

After you've made the final changes, please squash the commits into one commit. You can check https://github.com/mozilla/pdf.js/wiki/Squashing-Commits for how to do that easily. Thank you!

File: src/display/svg.js (hunk 3 above)
Comment (ahuglajbclajep) [reply src/display/svg.js]: The #10487 PDF uses the TilingPattern on the first page only.
Before implementing TilingPattern, paths that should be drawn in red were black with `fill=""`.
Looking at the output of this PR implementation, the rendering of `<svg:pattern>` using `<svg:g clip-path="...">` is broken.
I can infer from this that there might be a mistake in the implementation associated with clip, but I didn't know exactly what was wrong.

File: src/display/svg.js [hunk 4]
```diff
@@ -982,12 +982,65 @@ SVGGraphics = class SVGGraphics {
...
+      this.current.fillColor = cssColor;
+      this.current.strokeColor = cssColor;
+    }
+    this.executeOpTree(this.convertOpList(operatorList));
+
+    // Restore saved state.
+    this.svg = svg;
+    this.transformMatrix = transformMatrix;
+    this.fillColor = fillColor;
+    this.strokeColor = strokeColor;
```
Comment (janpe2) [inline src/display/svg.js]: When you restore the state, the lines should be
```
this.current.fillColor = fillColor;
this.current.strokeColor = strokeColor;
```

**Comment by ahuglajbclajep** [comment]
@timvandermeij @janpe2 Please review it.

File: src/display/svg.js (hunk 3 above)
Comment (timvandermeij) [reply src/display/svg.js]: I have been thinking about this some more and I think it's best to leave fixing this for a follow-up. The PDF file already wasn't rendering correctly and there is an open issue for it. Given that this patch improves the other PDF files, I don't see this as a blocker for landing this.

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Nice work; thank you for improving the SVG back-end!

**Pull request merged**